### PR TITLE
[GSoC] Implement Corrected Commit Date

### PR DESCRIPTION
**Pull request by abhishekkumar2718**
This patch series implements the corrected commit date offsets as generation number v2, along with other pre-requisites.

Git uses topological levels in the commit-graph file for commit-graph traversal operations like 'git log --graph'. Unfortunately, topological levels can perform worse than committer date when parents of a commit differ greatly in generation numbers [1]. For example, 'git merge-base v4.8 v4.9' on the Linux repository walks 635,579 commits using topological levels and walks 167,468 using committer date. Since 091f4cf3 (commit: don't use generation numbers if not needed, 2018-08-30), 'git merge-base' uses committer date heuristic unless there is a cutoff because of the performance hit.

[1] https://lore.kernel.org/git/efa3720fb40638e5d61c6130b55e3348d8e4339e.1535633886.git.gitgitgadget@gmail.com/

Thus, the need for generation number v2 was born. As Git used to die when graph version understood by it and in the commit-graph file are different [2], we needed a way to distinguish between the old and new generation number without incrementing the graph version.

[2] https://lore.kernel.org/git/87a7gdspo4.fsf@evledraar.gmail.com/

The following candidates were proposed (https://github.com/derrickstolee/gen-test, https://github.com/abhishekkumar2718/git/pull/1):
- (Epoch, Date) Pairs.
- Maximum Generation Numbers.
- Corrected Commit Date.
- FELINE Index.
- Corrected Commit Date with Monotonically Increasing Offsets.

Based on performance, local computability, and immutability (along with the introduction of an additional commit-graph chunk which relieved the requirement of backwards-compatibility) Corrected Commit Date was chosen as generation number v2 and is defined as follows:

For a commit C, let its corrected commit date  be the maximum of the commit date of C and the corrected commit dates of its parents plus 1. Then corrected commit date offset is the difference between corrected commit date of C and commit date of C. As a special case, a root commit with the timestamp zero has corrected commit date of 1 to distinguish it from GENERATION_NUMBER_ZERO (that is, an uncomputed generation number).

While it was proposed initially to store corrected commit date offsets within Commit Data Chunk, storing the offsets in a new chunk did not affect the performance measurably. The new chunk is "Generation DATa (GDAT) chunk" and it stores corrected commit date offsets while CDAT chunk stores topological level. The old versions of Git would ignore GDAT chunk, using topological levels from CDAT chunk. In contrast, new versions of Git would use corrected commit dates, falling back to topological level if the generation data chunk is absent in the commit-graph file.

While storing corrected commit date offsets saves us 4 bytes per commit (as compared with storing corrected commit dates directly), it's however possible for the offset to overflow the space allocated. To handle such cases, we introduce a new chunk, _Generation Data Overflow_ (GDOV) that stores the corrected commit date. For overflowing offsets, we set MSB and store the position into the GDOV chunk, in a mechanism similar to the Extra Edges list chunk.

For mixed generation number environment (for example new Git on the command line, old Git used by GUI client), we can encounter a mixed-chain commit-graph (a commit-graph chain where some of split commit-graph files have GDAT chunk and others do not). As backward compatibility is one of the goals, we can define the following behavior:

While reading a mixed-chain commit-graph version, we fall back on topological levels as corrected commit dates and topological levels cannot be compared directly.

When adding new layer to the split commit-graph file, and when merging some or all layers (replacing them in the latter case), the new layer will have GDAT chunk if and only if in the final result there would be no layer without GDAT chunk just below it.

Thanks to Dr. Stolee, Dr. Narębski, Taylor Blau and SZEDER Gábor for their reviews.

I look forward to everyone's reviews!

Thanks
- Abhishek

-----

Improvements left for a future series:
- Save commits with generation data overflow and extra edge commits instead of looping over all commits. cf. <858sbel67n.fsf@gmail.com>
- Verify both topological levels and corrected commit dates when present. cf. <85pn4tnk8u.fsf@gmail.com>

Changes in version 7:
- Moved the documentation patch ahead of "commit-graph: implement corrected commit date" and elaborated on the introduction of generation number v2.

Changes in version 6:
- Fixed typos in commit message for "commit-graph: implement corrected commit date".
- Removed an unnecessary else-block in "commit-graph: implement corrected commit date".
- Validate mixed generation chain correctly while writing in "commit-graph: use generation v2 only if the entire chain does".
- Die if the GDAT chunk indicates data has overflown but there are is no generation data overflow chunk.

Changes in version 5:
- Explained a possible reason for no change in performance for "commit-graph: fix regression when computing bloom-filters"
- Clarified about the addition of a new test for 11-digit octal implementations of ustar.
- Fixed duplicate test names in "commit-graph: consolidate fill_commit_graph_info".
- Swapped the order "commit-graph: return 64-bit generation number", "commit-graph: add a slab to store topological levels" to minimize lines changed.
- Fixed the mismerge in "commit-graph: return 64-bit generation number"
- Clarified the preparatory steps are for the larger goal of implementing generation number v2 in "commit-graph: return 64-bit generation number".
- Moved the rename of "run_three_modes()" to "run_all_modes()" into a new patch "t6600-test-reach: generalize *_three_modes".
- Explained and removed the checks for `GENERATION_NUMBER_INFINITY` that can never be true in "commit-graph: add a slab to store topological levels".
- Fixed incorrect logic for verifying commit-graph in "commit-graph: implement corrected commit date".
- Added minor improvements to commit message of "commit-graph: implement generation data chunk".
- Added '--date <date>' option to test_commit() in 'test-lib-functions.sh' in "commit-graph: implement generation data chunk".
- Improved coding style (also in tests) for "commit-graph: use generation v2 only if entire chain does".
- Simplified test repository structure in "commit-graph: use generation v2 only if entire chain does" as only the number of commits in a split commit-graph layer are relevant.
- Added a new test in "commit-graph: use generation v2 only if entire chain does" to check if the layers are merged correctly.
- Explicitly mentioned commit "091f4cf3" in the commit-message of "commit-graph: use corrected commit dates in paint_down_to_common()".
- Minor corrections to documentation in "doc: add corrected commit date info".
- Minor corrections to coding style.

Changes in version 4:
- Added GDOV to handle overflows in generation data.
- Added a test for writing tip graph for a generation number v2 graph chain in t5324-split-commit-graph.sh
- Added a section on how mixed generation number chains are handled in `Documentation/technical/commit-graph-format.txt`
- Reverted unimportant whitespace, style changes in commit-graph.c
- Added header comments about the order of comparision for  compare_commits_by_gen_then_commit_date in commit.h, compare_commits_by_gen in commit-graph.h
- Elaborated on why t6404 fails with corrected commit date and must be run with GIT_TEST_COMMIT_GRAPH=1in the commit "commit-reach: use corrected commit dates in paint_down_to_common()"
- Elaborated on write behavior for mixed generation number chains in the commit "commit-graph: use generation v2 only if entire chain does"
- Added notes about adding the topo_level slab to struct write_commit_graph_context as well as struct commit_graph.
- Clarified commit message for "commit-graph: consolidate fill_commit_graph_info"
- Removed the claim "GDAT can store future generation numbers" because it hasn't been tested yet.

Changes in version 3:
- Reordered patches to implement corrected commit date before generation data chunk [3].
- Split "implement corrected commit date" into two patches - one introducing the topo level slab and other implementing corrected commit dates.
- Extended split-commit-graph tests to verify at the end of test.
- Use topological levels as generation number if any of split commit-graph files do not have generation data chunk.

[3] https://lore.kernel.org/git/aee0ae56-3395-6848-d573-27a318d72755@gmail.com/

Changes in version 2:
- Add tests for generation data chunk.
- Add an option `GIT_TEST_COMMIT_GRAPH_NO_GDAT` to control whether to write generation data chunk.
- Compare commits with corrected commit dates if present in paint_down_to_common().
- Update technical documentation.
- Handle mixed generation commit chains.
- Improve commit messages for "commit-graph: fix regression when computing bloom filter", "commit-graph: consolidate fill_commit_graph_info",
- Revert unnecessary whitespace changes.
- Split uint_32 -> timestamp_t change into a new commit.

cc: Derrick Stolee <stolee@gmail.com>, Jakub Narębski <jnareb@gmail.com>
cc: Abhishek Kumar <abhishekkumar8222@gmail.com>
cc: SZEDER Gábor <szeder.dev@gmail.com>
cc: Taylor Blau <me@ttaylorr.com>

### CODE REVIEW

**Comment by derrickstolee** [comment]
@abhishekkumar2718 thanks for opening this so I can see your progress! Please @-mention me when you have working builds so I can take a  quick review before you submit.

**Comment by derrickstolee** [comment]
> User abhishekkumar2718 is now allowed to use GitGitGadget.
> 
> WARNING: abhishekkumar2718 has no public email address set on GitHub

@abhishekkumar2718 please fix the public email address so GitGitGadget can CC you properly.

File: commit.h [hunk 1]
```diff
@@ -15,6 +15,8 @@
 #define GENERATION_NUMBER_MAX 0x3FFFFFFF
 #define GENERATION_NUMBER_ZERO 0
 
+#define GENERATION_NUMBER_V2_INFINITY 0x7FFFFFFFF
```
[the remark this answers is not in the document]
Comment (derrickstolee) [reply commit.h:18]: I can see why you went this way, but it's due to a later commit implementing generation number v2 in a non-obvious way.

File: commit-graph.c [hunk 2]
```diff
@@ -109,8 +109,14 @@ timestamp_t commit_graph_generation(const struct commit *c)
...
 		return GENERATION_NUMBER_V2_INFINITY;
+	else if (data->odate)
+		/*
+		 * MSB differentiates between zero offset and missing
+		 * generation data chunk
+		 */
+		return c->date + (data->odate & 0x7FFFFFFF);
 
-	return data->generation;
+	return data->level;
```
Comment (derrickstolee) [inline commit-graph.c]: I think `data` should just have a `uint64_t generation`, and you fill that value with the v2 generation number (`date + offset`) when parsing the commit OR fill it with the v1 generation number if `GDAT` doesn't exist (again, when parsing the commit). That makes all of your other edits unnecessary, as we can just compare the `generation` directly and know we are correct one way or another.

File: commit-graph.c [hunk 3]
```diff
@@ -776,10 +781,16 @@ static void fill_commit_graph_info(struct commit *item, struct commit_graph *g,
 	date_low = get_be32(commit_data + g->hash_len + 12);
 	item->date = (timestamp_t)((date_high << 32) | date_low);
 
+	/* Read topological level from CDAT */
+	graph_data->level = get_be32(commit_data + g->hash_len + 8) >> 2;
```
Comment (derrickstolee) [inline commit-graph.c:785]: I don't think we will want `level` at all if we have generation v2, right?

Comment (derrickstolee) [reply commit-graph.c:785]: But, if we are persisting it across the files, then I understand why you want to keep it in the struct.

**Comment by abhishekkumar2718** [comment]
@derrickstolee @ttaylorr

I am just unable to make much progress with writing tests. Here's where I am at now:
- We should control write behavior (i.e., control whether to write generation data chunk) rather than read behavior (i.e., control whether to read generation data chunk as Taylor suggested on the mailing list with `commitGraph.readsGenerationData`). 

The primary reason being that ignoring generation data chunk is not the same as reading a commit-graph without GDAT chunk. We possibly might have entry points that we do not guard with `if (g->read_generation_data)` condition.

- In the fifth patch (implement generation data chunk), 
  - Add a test `commit graph understands commitGraph.writeGenerationData` to write a graph with and without GDAT chunk (and use `graph_read_expect` to verify). 
  - Repeat `graph exists` without GDAT chunk to test normal git operations.
  - Repeat tests to corrupt graphs with GDAT chunk and without it.

- In the sixth patch (implement corrected commit date)
  - I drop the `detect incorrect generation number non-zero generation` for GDAT tests. To get a generation number zero with corrected commit date would require us to corrupt 35 bits (34 for commit date, 1 bits for corrected date offset), which is just unlikely. We still retain the test for the topological level.

WIll that be comprehensive enough?

Related to tests, as we replace `compare_commits_by_commit_date` in `paintdown_to_common()` with `compare_commits_by_gen`, the test `virtual trees were processed` fails in `t6024-recursive-merge.sh`.

The order commits merged changes from (A, B, and then A, C) to (A, C, and then A, B). While the change of order is correct, the commit oid for the intermediate commit changes between commit date (i.e testing with `GIT_TEST_COMMIT_GRAPH=0`) and corrected commit date (i.e testing with `GIT_TEST_COMMIT_GRAPH=1`).

How can I get the SHA256 OID for the new intermediate commit?

File: commit-graph.c [hunk 4]
```diff
@@ -1710,6 +1735,12 @@ static int write_commit_graph_file(struct write_commit_graph_context *ctx)
 	chunks[2].id = GRAPH_CHUNKID_DATA;
 	chunks[2].size = (hashsz + 16) * ctx->commits.nr;
 	chunks[2].write_fn = write_graph_chunk_data;
+	if (ctx->write_generation_data) {
```
Comment (derrickstolee) [inline commit-graph.c:1761]: This is absolutely the right way to structure this condition.

File: commit-graph.c [hunk 5]
```diff
@@ -2129,6 +2161,9 @@ int write_commit_graph(struct object_directory *odb,
 		}
 	}
 
+	git_config_get_bool("commitGraph.writeGenerationData", &write_generation_data);
+	ctx->write_generation_data = write_generation_data;
+
```
Comment (derrickstolee) [inline commit-graph.c:2190]: This doesn't need to be a config, as we don't really want users to toggle this. It could be a `GIT_TEST_COMMIT_GRAPH_NO_GDAT` environment variable. Then, your test could write with `GIT_TEST_COMMIT_GRAPH_NO_GDAT=1` and then check all of the reading cases.

A good way to check that everything still works properly would be to extend `t/t6600-test-reach.sh` to write one commit-graph file in this mode (say, `commit-graph-no-gdat` to match the `commit-graph-[full|half]` that already exist), then update `run_three_modes` to be `test_all_modes` and run one test using that `-no-gdat` file.

File: t/t5324-split-commit-graph.sh [hunk 6]
```diff
@@ -425,4 +425,54 @@ done <<\EOF
 0600 -r--------
 EOF
 
+test_expect_success 'setup repo for mixed generation commit-graph-chain' '
```
Comment (abhishekkumar2718) [inline t/t5324-split-commit-graph.sh:428]: @derrickstolee @jnareb @ttaylorr

While writing test for split commit-graph, I wondered about the following scenario:

We start off with New Git and create a base commit-graph. Then we switch to old Git (possibly installed with an IDE, for example) and add some more commits. If we try to write to a new commit-graph file, we wind up with a chain of commit-graph files with and without GDAT chunks.

In such a case, I have disabled generation numbers and print out a warning.

We could have tried to salvage some of the performance boost by using topological level. I am thinking of setting up adding another flag to struct commit_graph to indicate whether to read topological level or corrected commit date. Will that be worth the trouble?

Comment (derrickstolee) [reply t/t5324-split-commit-graph.sh:428]: When we are writing a split commit-graph, we should not add a GDAT chunk if the base commit graph doesn't have one. This will prevent "holes" in the middle of a chain, as you described. It means we won't get a GDAT chunk on top of old commit-graphs unless we merge all of the levels.

With this in mind, it would be good to write a commit-graph without a GDAT chunk (using `GIT_TEST_COMMIT_GRAPH_NO_GDAT=1`), then add a new layer on top (without that env var) and see that the GDAT chunk doesn't exist there. Finally, rewrite the whole commit-graph using `--split=replace` and see that the single layer _does_ have a GDAT chunk.

Does that make sense?

**Comment by bk2204** [comment]
> How can I get the SHA256 OID for the new intermediate commit?

The easiest way to extract an object ID for SHA-256 is to just compute it if you can, using something like this:

```sh
oid=$(git rev-parse --verify HEAD^)
```

Unless your test thinks it's very important to get exactly the specific commit ID, computing it is much easier and makes the test easier to change in the future.  If that doesn't work, what I generally do is run the test with `GIT_TEST_DEFAULT_HASH=sha256` and then insert a debugging command to print the object I'm interested in.  That will require that you temporarily merge with or rebase on `seen`, since that's where that code is right now.

File: t/t5324-split-commit-graph.sh (hunk 6 above)
Comment (jnareb) [reply t/t5324-split-commit-graph.sh:428]: I agree with @derrickstolee that when writing section of a split commit-graph we should not add a GDAT chunk if the previous file in the chain doesn't have it -- this solution simplifies reasoning (and should be only temporary measure until all Git is upgraded to generation number v2 capable version).

I'm not sure if another flag in `struct commit_graph` would be necessary, though it might be the best solution anyway. If the topmost commit we start from, or the top file in the commit-graph chain doesn't have generation number v2 (no GDAT chunk), we switch to generation number v1. Otherwise we can safely use generation number v2.

File: commit-graph.c (hunk 5 above)
Comment (abhishekkumar2718) [reply commit-graph.c:2190]: Wow! That's so simple.

I spent a week wondering all the ways I should test reachability without GDAT chunk and the answer was as simple as to leverage the hard work that done by you and other contributers!

Thanks for the detailed response.

**Comment by abhishekkumar2718** [comment]
@derrickstolee @jnareb 

> Thanks for reaching out, @abhishekkumar2718. We are here to help when you have issues.
> 
> (I make some pointers to how to exclude the GDAT chunk and test the compatibility mode in code comments.)

Thanks for the speedy reply. With your pointers and encouragement, I made a lot more progress. Fixing some tests with split-commit-graph but we should be ready to publish the second version by tomorrow night.

> I think we shouldn't worry about the corruption tests right now. Let's revisit that at another time. Keep the topological level test for now.
> 
> This is a known issue with that test. The merge-base is not well-defined in that case, so the test can flip between the two. This was already modified slightly in [859fdc0](https://github.com/gitgitgadget/git/commit/859fdc0c3cf9ad8cdd5eacaa24aee481bc1e7bc1) (commit-graph: define GIT_TEST_COMMIT_GRAPH, 2018-08-29) due to the generation numbers causing an issue. You might need to enable this config option in more places in that test script, or just at the beginning.

Right, as Brian mentioned it's easier to compute OID if it's not very important and I don't think it's important to match the exact OID. We care more about whether all commits are connected in a consistent way rather than the exact order of the merges.

File: t/t5324-split-commit-graph.sh (hunk 6 above)
Comment (abhishekkumar2718) [reply t/t5324-split-commit-graph.sh:428]: > When we are writing a split commit-graph, we should not add a GDAT chunk if the base commit graph doesn't have one. This will prevent "holes" in the middle of a chain, as you described. It means we won't get a GDAT chunk on top of old commit-graphs unless we merge all of the levels.
> 
> With this in mind, it would be good to write a commit-graph without a GDAT chunk (using `GIT_TEST_COMMIT_GRAPH_NO_GDAT=1`), then add a new layer on top (without that env var) and see that the GDAT chunk doesn't exist there. Finally, rewrite the whole commit-graph using `--split=replace` and see that the single layer _does_ have a GDAT chunk.
> 
> Does that make sense?

Yes, that makes complete sense to me. I can see why we shouldn't merge split commit-graph chain unless explicitly ordered by the user.

File: commit-graph.c [hunk 7]
```diff
@@ -381,6 +385,20 @@ struct commit_graph *parse_commit_graph(void *graph_map, size_t graph_size)
 				graph->chunk_commit_data = data + chunk_offset;
 			break;
 
+		case GRAPH_CHUNKID_GENERATION_DATA:
```
Comment (derrickstolee) [inline commit-graph.c:388]: (Commenting on the commit message here)

> We test the overflow-related code with the following repo history:
> 
>            F - N - U
>           /         \
> U - N - U            N
>          \          /
> 	  N - F - N
> 
> Where U is a commit with committer date of zero seconds since Unix
> epoch, N is a commit with committer date of 1112354055 (default
> committer date for the test suite) seconds since Unix epoch and F is
> a commit with committer date of (2 ^ 31 - 2) seconds since Unix epoch.

My nitpick here is that you use `F`, `N`, and `U` to indicate a property of the commits, but they are not "a commit". Instead, I would reword this as "The commits denoted by 'U' have committer date of zero seconds...".

File: commit-graph.c [hunk 8]
```diff
@@ -747,7 +765,18 @@ static void fill_commit_graph_info(struct commit *item, struct commit_graph *g,
 	date_low = get_be32(commit_data + g->hash_len + 12);
 	item->date = (timestamp_t)((date_high << 32) | date_low);
 
-	graph_data->generation = get_be32(commit_data + g->hash_len + 8) >> 2;
+	if (g->chunk_generation_data) {
+		offset = (timestamp_t) get_be32(g->chunk_generation_data + sizeof(uint32_t) * lex_index);
+
+		if (offset & CORRECTED_COMMIT_DATE_OFFSET_OVERFLOW) {
+			offset_pos = offset - CORRECTED_COMMIT_DATE_OFFSET_OVERFLOW;
```
Comment (derrickstolee) [inline commit-graph.c:803]: I wonder if this would be clearer using bitwise XOR instead of subtraction:

`offset_pos = offset ^ CORRECTED_COMMIT_DATE_OFFSET_OVERFLOW;`

File: commit-graph.c [hunk 9]
```diff
@@ -747,7 +765,18 @@ static void fill_commit_graph_info(struct commit *item, struct commit_graph *g,
...
+			graph_data->generation = get_be64(g->chunk_generation_data_overflow + sizeof(timestamp_t) * offset_pos);
```
Comment (derrickstolee) [inline commit-graph.c:804]: Is `timestamp_t` always 64-bits? Perhaps we should use a constant or `sizeof(uint64_t)` here?

File: commit-graph.c [hunk 10]
```diff
@@ -747,7 +765,18 @@ static void fill_commit_graph_info(struct commit *item, struct commit_graph *g,
...
+		}
+		else
+			graph_data->generation = item->date + offset;
+	}
+	else
+		graph_data->generation = get_be32(commit_data + g->hash_len + 8) >> 2;
```
Comment (derrickstolee) [inline commit-graph.c:810]:
```
} else
```

Comment (derrickstolee) [reply commit-graph.c:810]: for the `else` above, too.

File: commit-graph.c [hunk 11]
```diff
@@ -1098,8 +1129,50 @@ static int write_graph_chunk_data(struct hashfile *f,
...
+	for (i = 0; i < ctx->commits.nr; i++) {
+		struct commit *c = ctx->commits.list[i];
+		timestamp_t offset = commit_graph_data_at(c)->generation - c->date;
+		display_progress(ctx->progress, ++ctx->progress_cnt);
+
+		if (offset > GENERATION_NUMBER_V2_OFFSET_MAX)
+		{
+			offset = CORRECTED_COMMIT_DATE_OFFSET_OVERFLOW | num_generation_data_overflows;
+			num_generation_data_overflows++;
+		}
```
Comment (derrickstolee) [inline commit-graph.c:1176]: `if (...) {`

File: commit-graph.c [hunk 12]
```diff
@@ -1098,8 +1129,50 @@ static int write_graph_chunk_data(struct hashfile *f,
...
+{
+	int i;
+	for (i = 0; i < ctx->commits.nr; i++) {
+		uint32_t packedOffset[2];
+		struct commit *c = ctx->commits.list[i];
+		timestamp_t offset = commit_graph_data_at(c)->generation - c->date;
+		display_progress(ctx->progress, ++ctx->progress_cnt);
+
+		if (offset > GENERATION_NUMBER_V2_OFFSET_MAX)
+		{
```
Comment (derrickstolee) [inline commit-graph.c:1195]: `if (..) {`

File: commit-graph.c [hunk 13]
```diff
@@ -1098,8 +1129,50 @@ static int write_graph_chunk_data(struct hashfile *f,
...
+			packedOffset[0] = htonl((offset >> 32));
+			packedOffset[1] = htonl(offset);
+			hashwrite(f, packedOffset, 8);
```
Comment (derrickstolee) [inline commit-graph.c:1198]: I believe @ttaylorr has a `hashwrite_be64()` in the works for a related issue, but here I would consider instead (while that is in-progress):

```
hashwrite_be32(f, offset >> 32);
hashwrite_be32(f, (uint32_t)offset);
```

File: commit-graph.c [hunk 14]
```diff
@@ -1098,8 +1129,50 @@ static int write_graph_chunk_data(struct hashfile *f,
...
+		}
+	}
+
+	return 0;
+}
+
 static int write_graph_chunk_extra_edges(struct hashfile *f,
-					 struct write_commit_graph_context *ctx)
+					  struct write_commit_graph_context *ctx)
```
Comment (derrickstolee) [inline commit-graph.c:1206]: Let's keep this out of this patch, as we have enough going on.

**Review by derrickstolee — COMMENTED** [review]
Sorry that I've been absent from this thread. I think this approach looks good and all of my notes are minor.

File: commit-graph.c (hunk 9 above)
Comment (abhishekkumar2718) [reply commit-graph.c:821]: `timestamp_t` is not always 64-bits, it's implementation specific. Since the neighbouring code prefers to use constant (for example: `date_high = get_be32(commit_data + g->hash_len + 8) 0x3)`, I will move ahead with using constant.

**Comment by dscho** [comment]
@abhishekkumar2718 the branch already hit `next`, therefore the Git maintainer prefers add-on patches from now on, not re-written commits (except in serious cases where the series has to be kicked out of `next`).

**Comment by dscho** [comment]
> /submit
> 

Ummm. @abhishekkumar2718 did you miss what I wrote about `next`?
